### ISSUE TITLE
ardb larger in memory than expected?

### ISSUE THREAD

**Issue by bpytlik** [yinqiwen/ardb#126]
I've got an ardb database with about 5700 keys in it but it seems to be consuming about 773m of RAM. Here's the top output:
```
  PID USER      PR  NI  VIRT  RES  SHR S %CPU %MEM    TIME+  COMMAND           
24421 root      20   0 1900m 773m 1260 S    0 38.2  26:07.82 ardb-server  
```

That works out to about 135k per key. While our values are large JSON blobs, I don't think they're THAT big. We're using leveldb with these settings:
```
#leveldb's options , 0 means default setting
leveldb.block_cache_size       512m
leveldb.block_size             0
leveldb.write_buffer_size 64m
leveldb.block_restart_interval 0
leveldb.max_open_files 50
leveldb.bloom_bits             10
leveldb.batch_commit_watermark 1024
# Specify the leveldb's compression
# This can be one of snappy/none
leveldb.compression            snappy
```

Based on the documentation in the README, I was expecting a max memory size of around 300M, so I'm bit confused as to what's going on.
```
  write_buffer_size*2    
 + max_open_files*4194304    
 + (size in NewLRUCache initialization)  
```
so that's
```
64m * 2 + 50 * 4194304 + 0 = 2 * 64 * 1024 * 1024 + 50 * 4194304 = 343932928 bytes
```
So that's about 328M.

I don't know for sure, but I think the memory usage may jump when the database is compacted. This is anecdotal but in all the cases where I've seen memory usage outside of my expectations, I've seen a compaction message like this in the logs:
```
2014-11-03_16:48:48.78576 [24421] 11-03 16:48:48,785 INFO Start compact DB.
2014-11-03_16:48:54.63444 [24421] 11-03 16:48:54,634 INFO Cost 5849ms to compact DB.
```

I restarted the process to load in the same data and here's the new top output (I waited until the "ready to accept connections" message appeared in the output to take a snapshot of the size):
```
 3897 root      20   0 1116m 6952 2228 S    0  0.3   0:02.55 ardb-server 
```

So just by restarting the process, the memory usage dropped from 773m to 7m...

Is this expected? Is there anything we can do to avoid the large memory consumption?

**Comment by bpytlik** [comment]
I thought this had fixed my issues but I've recently run into ardb taking over 1G of memory when it's supposed to only take around 800M by my calculation. 

Could you elaborate on this fix: 22c6816cd9c2f346d4a582b3093b515b1313d2c2 ?
I'm curious how severe of a memory leak this is/was. I wonder if that's the source of our issues (since we're running a version prior to that version). Can you characterize how much was leaked per redis connection or some other metric?


We are using leveldb as our backing store. Here are it's settings:
```
leveldb.block_cache_size 64m
leveldb.block_size             0
leveldb.write_buffer_size 64m
leveldb.block_restart_interval 0
leveldb.max_open_files 50
leveldb.bloom_bits             10
leveldb.batch_commit_watermark 1024
# Specify the leveldb's compression
# This can be one of snappy/none
leveldb.compression            snappy
```

Here are the buffer limits I have in place:
```
slave-client-output-buffer-limit 256mb
pubsub-client-output-buffer-limit 32mb
```


I have the L1 cache off:
```
L1-zset-read-fill-cache   no
L1-zset-seek-load-cache   no
L1-set-read-fill-cache    no
L1-set-seek-load-cache    no
L1-hash-read-fill-cache   no
L1-hash-seek-load-cache   no
L1-list-read-fill-cache   no
L1-list-seek-load-cache   no
L1-string-read-fill-cache no
```

Do I need to change numbers associated with the L1 cache, like "hash-max-ziplist-entries" to prevent that from taking up space even if it's not being used?

We are using this commit version of ardb: e5d9f7efa2d4c821768380411bfa0e7ee77bc69f.

Is there something else we need to manipulate to limit the amount of memory that ardb takes?

Here's the diff with the default config file:
```
4c4
< home  ..
---
> home /shared/stats
27c27
< listen             0.0.0.0:16379
---
> listen 127.0.0.1:6382
34,35d33
< #listen on another address with specified thread-pool-size & qps-limit
< #listen             0.0.0.0:36380
39,40d36
< #listen on unix socket
< #listen             /tmp/ardb.sock
48c44
< leveldb.block_cache_size       512m
---
> leveldb.block_cache_size 64m
50c46
< leveldb.write_buffer_size      128m
---
> leveldb.write_buffer_size 64m
52c48
< leveldb.max_open_files         1024
---
> leveldb.max_open_files 50
245,246c241
< # repl-backlog-size 500m 
< repl-backlog-size           1G
---
> repl-backlog-size 50m
482a478
> 
```

Thanks

**Comment by yinqiwen** [comment]
Comment on https://github.com/yinqiwen/ardb/commit/22c6816cd9c2f346d4a582b3093b515b1313d2c2

If you set 'L1_*-fill-cache' to no, the L1 cache is disable too.

Can you disable the auto compact task by setting `compact-enable` to no？

**Comment by bpytlik** [comment]
I can try that but given that there's a high turnover of data in our DB (ie, we add and delete things a lot), I'd be concerned about the impact that'd have on disk space. Also, given the information below, I think that's an unlikely source of the problem:

What I can say is that, assuming I'm reading the logs correctly in that:
```
2015-01-14_15:25:58.10080 [4439] 01-14 15:25:58,100 INFO Start compact DB.
```
means that auto compaction has started and
```
2015-01-14_15:26:28.62878 [4439] 01-14 15:26:28,628 INFO Cost 30528ms to compact DB.
```
means that compaction has finished, then we never hit the OOM killer while doing a compact and we successfully compacted the DB many times w/out triggering the issue. Also does not seem to be happening immediately after a compact has finished, or after a particular number of compacts. I also see several examples where we exhausted memory before we did any compacts.

If after that info, you'd still like me to try disabling auto compact, let me know and I'll give it a shot sometime this week.

**Comment by yinqiwen** [comment]
It seems that the compact task would not consume more memory, there is no need to disable  auto compact.
And in your configuration, `leveldb.block_cache_size` is configured by 512M, this is just the LRU cache of leveldb, which would consume at most 512MB. 

Since there is no way to estimate memory usage by ardb&eveldb,  there is a list which may consume many rss memory.
1. Leveldb memory usage (64m * 2 + 50 * 4194304 + 512MB = 2 * 64 * 1024 * 1024 + 50 * 4194304 = 343932928 bytes) = 840MB
2. Replication log (50MB)
3. L1 cache (already disabled)
4. Network (hard to calculate, depends on how many clients connected, in most time ,it would consume less than 50MB)
5. Others

**Comment by bpytlik** [comment]
We're seeing memory usage in the 1.4G range when we hit this issue. I agree with your above analysis. We have on the order of 10-20 clients connected I think. Certainly less than 100 afaik. Is it possible the leak fixed in the commit mentioned above could leak 400-500m over 1-2 days?

**Comment by yinqiwen** [comment]
I'm not sure the fix is the root cause of your issue. 
The commit https://github.com/yinqiwen/ardb/commit/22c6816cd9c2f346d4a582b3093b515b1313d2c2 only fix a memory leak when the reply is array of arrays, that's a very rare situation, most commands would only return string or array of strings.

**Comment by bpytlik** [comment]
Sorry, I took another look at things and realized that there was one problem w/ the analysis of the anticipated memory usage. We set leveldb.block_cache_size to be 64M not 512M. So that means our leveldb memory usage should be:
64m * 2 + 50 * 4194304 + 64MB = 2 * 64 * 1024 * 1024 + 50 * 4194304 + 64 * 1024 * 1024 = 411041792 bytes = 392MB

Let's say 100M for the network and replication log, that gives a total of 492M. Again, we were seeing usage at 1.4G when the process was killed.
